test/functional/test-impression.js

```diff
@@ -80,4 +86,66 @@ describe('impression', () => {
       });
     });
   });
+
+  it('should do nothing if response is not received', () => {
+    toggleExperiment(window, 'alp', true);
+    viewer.getParam.withArgs('click').returns('https://www.example.com');
+    xhr.fetchJson = () => {
+      setTimeout(() => {
+        return Promise.resolve({
+          'location': 'test_location?gclid=654321',
+        });
+      }, 5000);
+    };
+    const clock = sandbox.useFakeTimers();
+    const promise = new Promise(resolve => {
+      setTimeout(() => {
+        resolve();
+      }, 2000);
+    });
+    clock.tick(2001);
+    return promise.then(() => {
+      expect(window.location.href).to.not.contain('gclid=654321');
+      xhr.fetchJson = () => {
+        return Promise.resolve();
+      };
+    });
+  });
+
+  it('should do nothing if get empty response', () => {
+    toggleExperiment(window, 'alo', true);
+    viewer.getParam.withArgs('click').returns('https://www.example.com');
+    const prevHref = window.location.href;
+    maybeTrackImpression(window);
+    return Promise.resolve().then(() => {
+      return Promise.resolve().then(() => {
+        expect(window.location.href).to.equal(prevHref);
+        return getTrackImpressionPromise().should.be.fulfilled;
+      });
+    });
+  });
+
+  it('should replace location href only with query params', () => {
+    toggleExperiment(window, 'alp', true);
+    viewer.getParam.withArgs('click').returns('https://www.example.com');
+
+    xhr.fetchJson = () => {
+      return Promise.resolve({
+        'location': 'test_location?gclid=123456',
+      });
+    };
+    const prevHref = window.location.href;
+    maybeTrackImpression(window);
+    return Promise.resolve().then(() => {
+      return Promise.resolve().then(() => {
```

Comment:
This is to intentionally create some delay here before I run the test. Basicly to wait until `xhr.fetchJson` and `viewer.whenFirstVisible` to resolve.